homeassistant/components/tplink/light.py

```diff
@@ -377,21 +387,16 @@ def set_light_state(
         self, old_light_state: LightState, new_light_state: LightState
     ) -> None:
         """Set the light state."""
-        # Calling the API with the new state information.
-        if new_light_state.state != old_light_state.state:
-            if new_light_state.state:
-                self.smartbulb.state = SmartBulb.BULB_STATE_ON
-            else:
-                self.smartbulb.state = SmartBulb.BULB_STATE_OFF
-                return
-
-        if new_light_state.color_temp != old_light_state.color_temp:
-            self.smartbulb.color_temp = mired_to_kelvin(new_light_state.color_temp)
-
-        brightness_pct = brightness_to_percentage(new_light_state.brightness)
-        if new_light_state.hs != old_light_state.hs and len(new_light_state.hs) > 1:
-            hue, sat = new_light_state.hs
-            hsv = (int(hue), int(sat), brightness_pct)
-            self.smartbulb.hsv = hsv
-        elif new_light_state.brightness != old_light_state.brightness:
-            self.smartbulb.brightness = brightness_pct
+        old_state_param = old_light_state.to_param()
+        new_state_param = new_light_state.to_param()
+
+        diff = {
+            key: value
+            for key, value in new_state_param.items()
+            if new_state_param.get(key) != old_state_param.get(key)
+        }
+
```

Comment:
Also needs `color_temp` set to `0` if `hue` or `saturation` is set as it seems to block the setting
```suggestion
        if "hue" in diff or "saturation" in diff:
            diff["color_temp"] = 0
```